File: 122.best-time-to-buy-and-sell-stock-ii.cpp
/*
 * @lc app=leetcode id=122 lang=cpp
 *
 * [122] Best Time to Buy and Sell Stock II
 */
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int size = prices.size();
        if(size < 2) return 0;
        int vallyIndex = 0;
        int maxProfit = 0;
        int i = 1;
        while(prices[i] <= prices[vallyIndex] && i < size-1) {
            vallyIndex = i;
            ++i;
        }
        for(i -= 1;i < size-1;++i) {
            if(prices[i] > prices[vallyIndex] && prices[i] >= prices[i+1]) {
                maxProfit = maxProfit + prices[i] - prices[vallyIndex];
                vallyIndex = i;
                ++i;
                while(prices[i] <= prices[vallyIndex] && i < size-1) {
                    vallyIndex = i;
                    ++i;
                }
                --i;
            }
        }
        if(vallyIndex < size - 1 && prices[size - 1] > prices[vallyIndex])
            maxProfit += prices[size - 1] - prices[vallyIndex];
        return maxProfit;
    }
};

File: 122.best-time-to-buy-and-sell-stock-ii.py
#
# @lc app=leetcode id=122 lang=python3
#
# [122] Best Time to Buy and Sell Stock II
#

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        size = len(prices)
        if size == 0:
            return 0
        min_price = prices[0]
        max_profit = 0
        i = 0
        while i < size:
            while i < size and prices[i] < min_price:
                min_price = prices[i]
                i = i + 1

            while i < size-1 and prices[i+1] >= prices[i]:
                i = i + 1
            max_profit = max_profit + prices[i] - min_price
            if i < size-1:
                min_price = prices[i+1]
            i = i + 1

        return max_profit
            
        
# @lc code=end

